hugr-core/src/extension/resolution/test.rs

```diff
@@ -0,0 +1,292 @@
+            &[],
+            &ExtensionRegistry::new([float_types::EXTENSION.to_owned()]),
+        )
+        .unwrap();
+    func.add_dataflow_op(
+        CallIndirect {
+            signature: Signature::new_endo(vec![float64_type()]),
+        },
+        vec![loaded_func, func_i0],
+    )
+    .unwrap();
+
+    // Add one of the custom ops.
+    func.add_dataflow_op(op_a, vec![func_i1]).unwrap();
+
+    // A nested dataflow region.
+    let mut dfg = func.dfg_builder_endo([(bool_t(), func_i1)]).unwrap();
+    let dfg_inputs = dfg.input_wires().collect_vec();
+    dfg.add_dataflow_op(op_b, dfg_inputs.clone()).unwrap();
+    dfg.finish_with_outputs(dfg_inputs).unwrap();
+
+    // A tag
+    func.add_dataflow_op(
+        Tag::new(0, vec![vec![bool_t()].into(), vec![int_type(4)].into()]),
+        vec![func_i1],
+    )
+    .unwrap();
+
+    // Dfg control flow: Tail loop
+    let mut tail_loop = func
+        .tail_loop_builder([(bool_t(), func_i1)], [], vec![].into())
+        .unwrap();
+    let tl_inputs = tail_loop.input_wires().collect_vec();
+    tail_loop.add_dataflow_op(op_c, tl_inputs).unwrap();
+    let tl_tag = tail_loop.add_load_const(Value::true_val());
+    let tl_tag = tail_loop
+        .add_dataflow_op(
+            Tag::new(0, vec![vec![Type::new_unit_sum(2)].into(), vec![].into()]),
+            vec![tl_tag],
+        )
+        .unwrap()
+        .out_wire(0);
+    tail_loop.finish_with_outputs(tl_tag, vec![]).unwrap();
+
+    // Dfg control flow: Conditionals
+    let cond_tag = func.add_load_const(Value::unary_unit_sum());
+    let mut cond = func
+        .conditional_builder(([type_row![]], cond_tag), [], type_row![])
+        .unwrap();
+    let mut case = cond.case_builder(0).unwrap();
+    case.add_dataflow_op(op_e, [func_i1]).unwrap();
+    case.finish_with_outputs([]).unwrap();
+
+    // Cfg control flow.
+    let mut cfg = func
+        .cfg_builder([(bool_t(), func_i1)], vec![].into())
+        .unwrap();
+    let mut cfg_entry = cfg.entry_builder([type_row![]], type_row![]).unwrap();
+    let [cfg_i0] = cfg_entry.input_wires_arr();
+    cfg_entry.add_dataflow_op(op_d, [cfg_i0]).unwrap();
+    let cfg_tag = cfg_entry.add_load_const(Value::unary_unit_sum());
+    let cfg_entry_wire = cfg_entry.finish_with_outputs(cfg_tag, []).unwrap();
+    let cfg_exit = cfg.exit_block();
+    cfg.branch(&cfg_entry_wire, 0, &cfg_exit).unwrap();
+
+    // --------------------------------------------------
+
+    // Finally, finish the hugr and ensure it's using the right extensions.
+    func.finish_with_outputs(vec![]).unwrap();
+    let mut hugr = module
+        .finish_hugr(&build_extensions)
+        .unwrap_or_else(|e| panic!("{e}"));
+
+    // Check that the read-only methods collect the same extensions.
+    let mut collected_exts = ExtensionRegistry::default();
+    for node in hugr.nodes() {
+        let op = hugr.get_optype(node);
+        collected_exts.extend(collect_op_extensions(Some(node), op).unwrap());
```

Comment:
If you changed OpType::used_extensions to take the `Option<Node>` param as extra, which seems reasonable, then you could call that here and remove all direct calls to `collect_op_`(`type_`)`extensions`. Then hide those (or inline into the single caller `OpType::used_extensions`).

Reply:
I was calling the resolve methods explicitly, but it's exactly the same as `op.used_exts`.
I just changed it to that.